test/e2e/functional_test.go

```diff
@@ -190,35 +189,48 @@ func (s *FunctionalSuite) TestBuiltinEventTimeExtractor() {
-	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 400*time.Second)
 	defer cancel()
 
-	// TODO - Figure out a better way to wait for watermark to propagate.
-	time.Sleep(time.Second * 2)
-	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgTwo)).WithHeader("X-Numaflow-Event-Time", timeNow))
-	time.Sleep(time.Second * 2)
-	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgThree)).WithHeader("X-Numaflow-Event-Time", timeNow))
-	time.Sleep(time.Second * 2)
-	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgFour)).WithHeader("X-Numaflow-Event-Time", timeNow))
-	time.Sleep(time.Second * 2)
-	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgFive)).WithHeader("X-Numaflow-Event-Time", timeNow))
-	time.Sleep(time.Second * 2)
-
-	wm, err := client.GetPipelineWatermarks(ctx, pipelineName)
-	assert.NoError(s.T(), err)
-	edgeWM := wm[0].Watermarks[0]
-	// Watermark propagation can delay, we consider the test as passed as long as the retrieved watermark matches one of the assigned event times.
-	assert.True(s.T(), edgeWM == time.Date(2021, 4, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 3, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 2, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli())
+	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgTwo)))
+	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgThree)))
+	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgFour)))
+	w.SendMessageTo(pipelineName, "in", NewHttpPostRequest().WithBody([]byte(testMsgFive)))
+
+wmLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				s.T().Log("test timed out")
+				assert.Fail(s.T(), "timed out")
+				break wmLoop
+			}
+		default:
+			wm, err := client.GetPipelineWatermarks(ctx, pipelineName)
+			edgeWM := wm[0].Watermarks[0]
+			if wm[0].Watermarks[0] != -1 {
+				assert.NoError(s.T(), err)
+				if err != nil {
+					assert.Fail(s.T(), err.Error())
+				}
+				// Watermark propagation can delay, we consider the test as passed as long as the retrieved watermark matches one of the assigned event times.
+				assert.True(s.T(), edgeWM == time.Date(2021, 5, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 4, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 3, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 2, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 1, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli() || edgeWM == time.Date(2021, 5, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli())
```

Comment:
There are two `edgeWM == time.Date(2021, 5, 18, 21, 54, 42, 123000000, time.UTC).UnixMilli()`, please remove one of them.